README.md
# R client to the Rx Studio APIs

This R package provides helper functions to use the [Rx Studio](https://rx.studio) platform right from R -- either for programmatic access (e.g. integrating data sources, PK/PD modeling and reporting), or just for a CLI interface instead of the web GUI provided at https://app.rx.studio

## Authentication

The `simulate` functions requires a free or licensed Rx Studio account to authenticate with the Rx Studio APIs. If you don't have an account yet, please visit https://rx.studio/demo.

To authenticate with your Rx Studio account, you need to generate a long-live access token at https://app.rx.studio/profile/preferences. Please store your token at a secure place, as it provides full access to your Rx Studio account -- including past simulations, and potentially PHI as well! Tokens can be secured by binding to source IP addresses. and unused tokens can (and should) be revoked.

To use an access token in your R session, you need to call `set_access_token` once.

## Simulations

All simulation functions tied to the related Rx Studio API endpoints start with the `simulate` prefix, then the drug, method and PK/PD target names. For example, `simulate_cefepime_mc_emp_fcmin` will use "Empiric therapy" for "Cefepime" targeting "Free minimum concentration".

Example call:

```r
simulate_cefepime_mc_emp_fcmin(
  PATID = "Anonymous",
  AGE = 65,
  HEIGHT = 175,
  WEIGHT = 75,
  GENDER = "Male",
  CREATININE = 1,
  MODEL = "Nicasio et al. (2009) - ICU",
  MIC = 1,
  CMIN = 6,
  LOADINGDOSE = "Yes",
  CRCLCAP = "No cap",
  REGIMENS = list(
    regimen(1000, 8, 1),
    regimen(1000, 12, 1),
    regimen(1000, 24, 1),
    regimen(1250, 8, 1.25),
    regimen(1250, 12, 1.25),
    regimen(1250, 24, 1.25),
    regimen(1500, 8, 1.5),
    regimen(1500, 12, 1.5),
    regimen(1500, 24, 1.5)
  )
)
```

## Helper functions

As the Rx Studio APIs sometimes requires somewhat complex data structures (e.g. when submitting a list of Dosing Regimens for simulations, or specifying Historical Records of a patient, such as past doses or lab results), the following helper functions have been provided to simplify the simulation calls:

- `regimen`
- `dose`
- `concentration`

## Contact

Please feel free to reach out to us at any time either by [email](mailto:team@rx.studio) or by opening a GitHub ticket!


NEWS.md
# RxStudio 2021-11-25 

Initial release. 

Affected simulation endpoints:

* phenytoin_bay_ncmin 

# RxStudio 2021-10-15 

Initial release. 

Affected simulation endpoints:

* voriconazole_bay_cmin_mic_ratio 
* voriconazole_bay_fauc_mic_ratio 
* voriconazole_cfr_cmin_mic_ratio 
* voriconazole_cfr_fauc_mic_ratio 
* voriconazole_emp_cmin_mic_ratio 
* voriconazole_emp_fauc_mic_ratio 
* voriconazole_pta_cmin_mic_ratio 
* voriconazole_pta_fauc_mic_ratio 

# RxStudio 2021-09-23 

Initial release. 

Affected simulation endpoints:

* phenytoin_loading_ncmin_24h 
* phenytoin_optloading_ncmin_24h 

# RxStudio 2021-08-25 

Fix doses in comparative plots. 

Affected simulation endpoints:

* piperacillin-tazobactam_bay_fauc_mic_ratio 
* piperacillin-tazobactam_bay_fcmin 
* piperacillin-tazobactam_bay_fcmin_mic_ratio 
* piperacillin-tazobactam_bay_ftime_above_mic 
* piperacillin-tazobactam_emp_fauc_mic_ratio 
* piperacillin-tazobactam_emp_fcmin 
* piperacillin-tazobactam_emp_fcmin_mic_ratio 
* piperacillin-tazobactam_emp_ftime_above_mic 

# RxStudio 2021-07-28 

Initial release. 

Affected simulation endpoints:

* meropenem_emp_ftime_above_mic 

# RxStudio 2021-06-30 

Initial release. 

Affected simulation endpoints:

* amikacin_bay_cmax_mic_ratio 
* amikacin_cfr_cmax_mic_ratio 
* amikacin_pta_cmax_mic_ratio 

# RxStudio 2021-06-04 

Initial release. 

Affected simulation endpoints:

* ceftriaxone_bay_fauc_mic_ratio 
* ceftriaxone_bay_fcmin 
* ceftriaxone_bay_fcmin_mic_ratio 
* ceftriaxone_bay_ftime_above_mic 
* ceftriaxone_cfr_fauc_mic_ratio 
* ceftriaxone_cfr_fcmin 
* ceftriaxone_cfr_fcmin_mic_ratio 
* ceftriaxone_cfr_ftime_above_mic 
* ceftriaxone_emp_fauc_mic_ratio 
* ceftriaxone_emp_fcmin 
* ceftriaxone_emp_fcmin_mic_ratio 
* ceftriaxone_emp_ftime_above_mic 
* ceftriaxone_pta_fauc_mic_ratio 
* ceftriaxone_pta_fcmin 
* ceftriaxone_pta_fcmin_mic_ratio 
* ceftriaxone_pta_ftime_above_mic 

# RxStudio 2021-05-18 

Add sparklines to visualize the individual patient PK parameters on the population distribution. 

Affected simulation endpoints:

* amikacin_bay_auc_mic_ratio 
* amikacin_bay_cmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_ftime_above_mic 
* gentamicin_bay_auc_mic_ratio 
* gentamicin_bay_cmin_mic_ratio 
* meropenem_bay_fauc_mic_ratio 
* meropenem_bay_fcmin 
* meropenem_bay_fcmin_mic_ratio 
* meropenem_bay_ftime_above_mic 
* piperacillin-tazobactam_bay_fauc_mic_ratio 
* piperacillin-tazobactam_bay_fcmin 
* piperacillin-tazobactam_bay_fcmin_mic_ratio 
* piperacillin-tazobactam_bay_ftime_above_mic 
* tobramycin_bay_auc_mic_ratio 
* tobramycin_bay_cmin_mic_ratio 
* vancomycin_bay_auc_mic_ratio 
* vancomycin_bay_cmin_mic_ratio 

# RxStudio 2021-05-14 

Update Goti et al. (2018) model to use the actual covariance matrix instead of relying on standard deviations estimated from the published bootstrap confidence intervals. 

Affected simulation endpoints:

* vancomycin_bay_auc_mic_ratio 
* vancomycin_bay_cmin_mic_ratio 
* vancomycin_cfr_auc_mic_ratio 
* vancomycin_cfr_cmin_mic_ratio 
* vancomycin_emp_auc_mic_ratio 
* vancomycin_emp_cmin 
* vancomycin_pta_auc_mic_ratio 
* vancomycin_pta_cmin_mic_ratio 

# RxStudio 2021-05-02 

Add population model to the simulated blood concentration plot. Visualize up to the next 5 recommended doses until steady-state reached. Compute and visualize the predicted target PK/PD values until steady-state reached. 

Affected simulation endpoints:

* amikacin_bay_auc_mic_ratio 
* amikacin_bay_cmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_ftime_above_mic 
* gentamicin_bay_auc_mic_ratio 
* gentamicin_bay_cmin_mic_ratio 
* meropenem_bay_fauc_mic_ratio 
* meropenem_bay_fcmin 
* meropenem_bay_fcmin_mic_ratio 
* meropenem_bay_ftime_above_mic 
* piperacillin-tazobactam_bay_fauc_mic_ratio 
* piperacillin-tazobactam_bay_fcmin 
* piperacillin-tazobactam_bay_fcmin_mic_ratio 
* piperacillin-tazobactam_bay_ftime_above_mic 
* tobramycin_bay_auc_mic_ratio 
* tobramycin_bay_cmin_mic_ratio 
* vancomycin_bay_auc_mic_ratio 
* vancomycin_bay_cmin_mic_ratio 

# RxStudio 2021-04-09 

Show results with warning for unreliable models instead of error. 

Affected simulation endpoints:

* amikacin_bay_auc_mic_ratio 
* amikacin_bay_cmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_ftime_above_mic 
* gentamicin_bay_auc_mic_ratio 
* gentamicin_bay_cmin_mic_ratio 
* meropenem_bay_fauc_mic_ratio 
* meropenem_bay_fcmin 
* meropenem_bay_fcmin_mic_ratio 
* meropenem_bay_ftime_above_mic 
* piperacillin-tazobactam_bay_fauc_mic_ratio 
* piperacillin-tazobactam_bay_fcmin 
* piperacillin-tazobactam_bay_fcmin_mic_ratio 
* piperacillin-tazobactam_bay_ftime_above_mic 
* tobramycin_bay_auc_mic_ratio 
* tobramycin_bay_cmin_mic_ratio 
* vancomycin_bay_auc_mic_ratio 
* vancomycin_bay_cmin_mic_ratio 

# RxStudio 2021-04-05 

Update user interface with new input field making it possible to add further PK/PD models. 

Affected simulation endpoints:

* amikacin_bay_auc_mic_ratio 
* amikacin_bay_auc_mic_ratio 
* amikacin_bay_auc_mic_ratio 
* amikacin_bay_cmin_mic_ratio 
* amikacin_bay_cmin_mic_ratio 
* amikacin_bay_cmin_mic_ratio 
* amikacin_cfr_auc_mic_ratio 
* amikacin_cfr_cmin_mic_ratio 
* amikacin_emp_auc_mic_ratio 
* amikacin_emp_cmin 
* amikacin_emp_cmin_cmax 
* amikacin_emp_cmin_mic_ratio 
* amikacin_pta_auc_mic_ratio 
* amikacin_pta_cmin 
* amikacin_pta_cmin_cmax 
* amikacin_pta_cmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_fcmin_mic_ratio 
* cefepime_bay_ftime_above_mic 
* cefepime_bay_ftime_above_mic 
* cefepime_bay_ftime_above_mic 
* cefepime_bay_ftime_above_mic 
* cefepime_bay_ftime_above_mic 
* cefepime_cfr_fcmin_mic_ratio 
* cefepime_cfr_fcmin_mic_ratio 
* cefepime_cfr_fcmin_mic_ratio 
* cefepime_cfr_ftime_above_mic 
* cefepime_cfr_ftime_above_mic 
* cefepime_cfr_ftime_above_mic 
* cefepime_emp_fauc_mic_ratio 
* cefepime_emp_fauc_mic_ratio 

R/simulate_cefepime_mc_emp_fcmin.R
#' Cefepime » Empiric » fCmin
#' 
#' Cefepime » Empiric therapy » Free minimum concentration
#' 
#' \strong{Drug}:
#' Cefepime
#' 
#' \strong{Method}:
#' Simulate concentrations for multiple dosing regimens and select the optimal one, with regard to the target pharmacodynamic index.
#' 
#' \strong{PK/PD target}:
#' Free minimum blood plasma concentration (mg/L).
#' 
#' @param PATID Patient Identifier. User-provided free text (such as patient id, name or alias) to identify related simulations. Must be provided as string.
#' @param AGE Age. Age of the patient in years. Must be provided as numeric (min. 18, max. 120 year).
#' @param HEIGHT Height. Height of the patient. Must be provided as numeric (min. 100, max. 250 cm).
#' @param WEIGHT Weight. Actual body weight of the patient. Must be provided as numeric (min. 20, max. 500 kg).
#' @param GENDER Sex. Patient's sex for clinical decision-making. Must be provided as string ('Male' or 'Female').
#' @param CREATININE Creatinine. Serum creatinine. Must be provided as numeric (min. 0.01, max. 15 mg/dL).
#' @param MODEL Model for population of interest. Pharmacokinetic model to be used for specific patient type during simulations. Must be provided as string ('Nicasio et al. (2009) - ICU' or 'Tam et al. (2003) - General ward').
#' @param MIC MIC. Minimum Inhibitory Concentration (MIC). Must be provided as numeric (min. 0.01, max. 1024 mg/L).
#' @param CMIN Minimum concentration target. The PK/PD target can be provided as minimum blood plasma concentration (Cmin). Must be provided as numeric (min. 0.1, max. 200 mg/L).
#' @param LOADINGDOSE Loading dose. Loading dose is desired or not. Must be provided as string ('No' or 'Yes').
#' @param CRCLCAP Capping Creatinine Clearance. Whether to use capping for creatinine clearance. Must be provided as string ('No cap', '120 ml/min', '130 ml/min', '140 ml/min' or '150 ml/min').
#' @param REGIMENS Dosing Regimens. List of dosing regimens to be used in simulating target attainment, from which the dosing regimen with the smallest absolute difference from the desired target will be automatically selected. Must be provided as list of 1-20 'REGIMEN' values. Use the \code{regimen} helper function to define the REGIMEN values.
#' 
#' @examples \dontrun{
#' simulate_cefepime_mc_emp_fcmin(PATID = "Anonymous", 
#'     AGE = 65, HEIGHT = 175, 
#'     WEIGHT = 75, GENDER = "Male", 
#'     CREATININE = 1, MODEL = "Nicasio et al. (2009) - ICU", 
#'     MIC = 1, CMIN = 17.5, 
#'     LOADINGDOSE = "No", 
#'     CRCLCAP = "No cap", 
#'     REGIMENS = list(list(
#'         DOSE = 750, INTERVAL = 8, 
#'         TINF = 1, set = "REGIMEN"), 
#'         list(DOSE = 750, 
#'             INTERVAL = 12, 
#'             TINF = 1, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 750, 
#'             INTERVAL = 24, 
#'             TINF = 1, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1000, 
#'             INTERVAL = 8, 
#'             TINF = 1, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1000, 
#'             INTERVAL = 12, 
#'             TINF = 1, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1000, 
#'             INTERVAL = 24, 
#'             TINF = 1, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1250, 
#'             INTERVAL = 8, 
#'             TINF = 1.25, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1250, 
#'             INTERVAL = 12, 
#'             TINF = 1.25, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1250, 
#'             INTERVAL = 24, 
#'             TINF = 1.25, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1500, 
#'             INTERVAL = 8, 
#'             TINF = 1.5, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1500, 
#'             INTERVAL = 12, 
#'             TINF = 1.5, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1500, 
#'             INTERVAL = 24, 
#'             TINF = 1.5, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1750, 
#'             INTERVAL = 8, 
#'             TINF = 1.75, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1750, 
#'             INTERVAL = 12, 
#'             TINF = 1.75, 
#'             set = "REGIMEN"), 
#'         list(DOSE = 1750, 
#'             INTERVAL = 24, 
#'             TINF = 1.75, 
#'             set = "REGIMEN")))
#' }
#' 
#' @references \itemize{
#'   \item Nicasio et al. (2009): Population Pharmacokinetics of High-Dose, Prolonged-Infusion Cefepime in Adult Critically Ill Patients with Ventilator-Associated Pneumonia. In. Antimicrobial Agents and Chemotherapy. https://pubmed.ncbi.nlm.nih.gov/19188394/
#'   \item Tam et al. (2003): Pharmacokinetics and Pharmacodynamics of Cefepime in Patients with Various Degrees of Renal Function. In. Antimicrobial Agents and Chemotherapy. https://pubmed.ncbi.nlm.nih.gov/12760858/
#' }
#' @export 
#' @importFrom checkmate assert_number assert_string assert_choice
simulate_cefepime_mc_emp_fcmin <- function(PATID, AGE, HEIGHT, WEIGHT, GENDER, CREATININE, MODEL, MIC, CMIN, LOADINGDOSE, CRCLCAP, REGIMENS) {
  ## check args
  assert_string(PATID)
  assert_number(AGE,
    lower = 18,
    upper = 120
  )
  assert_number(HEIGHT,
    lower = 100,
    upper = 250
  )
  assert_number(WEIGHT,
    lower = 20,
    upper = 500
  )
  assert_string(GENDER)
  assert_choice(GENDER, c("Male", "Female"))
  assert_number(CREATININE,
    lower = 0.01,
    upper = 15
  )
  assert_string(MODEL)
  assert_choice(MODEL, c("Nicasio et al. (2009) - ICU", "Tam et al. (2003) - General ward"))
  assert_number(MIC,
    lower = 0.01,
    upper = 1024
  )
  assert_number(CMIN,
    lower = 0.1,
    upper = 200
  )
  assert_string(LOADINGDOSE)
  assert_choice(LOADINGDOSE, c("No", "Yes"))
  assert_string(CRCLCAP)
  assert_choice(CRCLCAP, c("No cap", "120 ml/min", "130 ml/min", "140 ml/min", "150 ml/min"))
  ## API call
  simulate("cefepime-mc-emp-fcmin", PATID = PATID, AGE = AGE, HEIGHT = HEIGHT, WEIGHT = WEIGHT, GENDER = GENDER, CREATININE = CREATININE, MODEL = MODEL, MIC = MIC, CMIN = CMIN, LOADINGDOSE = LOADINGDOSE, CRCLCAP = CRCLCAP, REGIMENS = REGIMENS)
}
